Add basic types and helper methods for model exporting

caisq opened the pull request:
Towards https://github.com/tensorflow/tfjs/issues/13

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/tensorflow/tfjs-core/990)
<!-- Reviewable:end -->

caisq commented:
cc @adarob as this touches the same files as https://github.com/tensorflow/tfjs-core/pull/965

ericdnielsen commented on src/io/io_utils.ts:
```diff
@@ -0,0 +1,136 @@
+
+    if (tensor.dtype !== 'float32' && tensor.dtype !== 'int32' &&
+        tensor.dtype !== 'bool') {
+      throw new Error(`Unsupported dtype: ${tensor.dtype}`);
```
Can we include the name of the tensor here as well.

ericdnielsen commented on src/io/io_utils.ts:
```diff
@@ -0,0 +1,136 @@
+      bytes = numel;
+      value = tensor(new Uint8Array(buffer, offset, numel), shape, 'bool');
+    } else {
+      throw new Error(`Unsupported dtype: ${dtype}`);
```
As above, please include the name fields as well.

ericdnielsen commented on src/io/io_utils.ts:
```diff
@@ -0,0 +1,136 @@
+    } else if (x as any instanceof Uint8Array) {
+      totalByteLength += x.length;
+    } else {
+      throw new Error(`Unsupported TypedArray subtype: ${x.constructor.name}`);
```
Are there minification concerns here with constructor.name?

davidsoergel commented:
Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/io/io_utils.ts, line 3 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAtGPZC1pHoX2gjOSA4:-LAtGPZC1pHoX2gjOSA5:bndh7q0) ([raw file](https://github.com/tensorflow/tfjs-core/blob/fae4217686afd84402cd1acd9d65fb6a5e527cf0/src/io/io_utils.ts#L3)):*
> ```
> Inc
> ```
LLC

---

*[src/io/io_utils.ts, line 39 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAtHwlNBploTPWRhjNd:-LAtHwlNBploTPWRhjNe:b8z9dcc) ([raw file](https://github.com/tensorflow/tfjs-core/blob/fae4217686afd84402cd1acd9d65fb6a5e527cf0/src/io/io_utils.ts#L39)):*
> ```
> const tensor
> ```
This shadows the imported tensor() function; can you use a different name?

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/990)*
<!-- Sent from Reviewable.io -->

dsmilkov commented:
Review status: 5 of 6 files reviewed at latest revision, 16 unresolved discussions.

---

*[src/index.ts, line 36 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAtKBky2nHd3p0lKpMc:-LAtKBky2nHd3p0lKpMd:b-w1p4w8) ([raw file](https://github.com/tensorflow/tfjs-core/blob/fae4217686afd84402cd1acd9d65fb6a5e527cf0/src/index.ts#L36)):*
> ```XML
> export {decodeTensors, encodeTensors} from './io/io_utils';
> // tslint:disable-next-line:max-line-length
> export {IOHandler, LoadHandler, ModelArtifacts, SaveHandler, SaveResult, WeightsManifestConfig, WeightsManifestEntry} from './io/types';
> ```

just a note to double-check that it is necessary for all these types to be exposed to public api. Want to make sure we don't expose things that are meant to be implementation detail

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/990)*
<!-- Sent from Reviewable.io -->

dsmilkov commented:
Review status: 5 of 6 files reviewed at latest revision, 17 unresolved discussions.

---

*[src/io/types.ts, line 99 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAtLSYFEjPp38vpZY-x:-LAtLSYGBrm-SAQ_2wCu:b4u7xey) ([raw file](https://github.com/tensorflow/tfjs-core/blob/fae4217686afd84402cd1acd9d65fb6a5e527cf0/src/io/types.ts#L99)):*
> ```XML
>    * any). This is applicable only to server-based saving routes.
>    */
>   resposnes?: Response[];
> ```

s/resposnes/responses

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/990)*
<!-- Sent from Reviewable.io -->

pyu10055 commented:
Review status: 5 of 6 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[src/io/io_utils.ts, line 32 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAtNjVwADAwRE2CN-BK:-LAtNjVwADAwRE2CN-BL:bkl322l) ([raw file](https://github.com/tensorflow/tfjs-core/blob/fae4217686afd84402cd1acd9d65fb6a5e527cf0/src/io/io_utils.ts#L32)):*
> ```
> encodeTensors
> ```
want to point out this method will only work for encode weights, the activation tensors could have multiple outputs, they are identified by the name + index. maybe better to rename to encodeWeights and decodeWeights?

---

*[src/io/io_utils.ts, line 34 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAtOZyY1HPG_3st-NYK:-LAtOZyY1HPG_3st-NYL:b4qbm45) ([raw file](https://github.com/tensorflow/tfjs-core/blob/fae4217686afd84402cd1acd9d65fb6a5e527cf0/src/io/io_utils.ts#L34)):*
> ```XML
>  * @throws Error: on unsupported tensor `dtype`.
>  */
> export async function encodeTensors(tensors: NamedTensorMap):
> ```

Also, how different is this code from the loadWeights method, are there plan to merge those two? 
I also want to raise attention about the PR by Adam in converter repo that supports weight quantization.  

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/990)*
<!-- Sent from Reviewable.io -->

caisq commented:
Review status: 5 of 6 files reviewed at latest revision, 20 unresolved discussions.

---

*[src/index.ts, line 36 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAtKBky2nHd3p0lKpMc:-LAuqWtk0gVPKnLbHPYq:bh78ce2) ([raw file](https://github.com/tensorflow/tfjs-core/blob/52a7e4c00d30635a04da146a4652619455b0a10e/src/index.ts#L36)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

just a note to double-check that it is necessary for all these types to be exposed to public api. Want to make sure we don't expose things that are meant to be implementation detail
</blockquote></details>

Done double checking. These are all needed by implementation in tfjs-layer.

---

*[src/index.ts, line 37 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAtD_8VBCkzmZO4BKBS:-LAuqjc61kEmtq-8yCLU:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/52a7e4c00d30635a04da146a4652619455b0a10e/src/index.ts#L37)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

add a newline here to separate serialization from optimizers.
</blockquote></details>

Done.

---

*[src/io/io_utils.ts, line 3 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAtGPZC1pHoX2gjOSA4:-LAuqpuY3Gc8Ca2Fo7bh:b-wkig27) ([raw file](https://github.com/tensorflow/tfjs-core/blob/52a7e4c00d30635a04da146a4652619455b0a10e/src/io/io_utils.ts#L3)):*
<details><summary><i>Previously, davidsoergel (David Soergel) wrote…</i></summary><blockquote>

> ```
> Inc
> ```
LLC
</blockquote></details>

Done making this change throughout src/io/

---

*[src/io/io_utils.ts, line 18 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAtE-ME1G9F1e6lfb-_:-LAurWAP1HFyVwPXcmaG:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/52a7e4c00d30635a04da146a4652619455b0a10e/src/io/io_utils.ts#L18)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

since this is not a test, avoid importing directly from `index`.
</blockquote></details>

Done.

---

*[src/io/io_utils.ts, line 32 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAtNjVwADAwRE2CN-BK:-LAutlY8AjKSIcBZRBR1:b-9zsj) ([raw file](https://github.com/tensorflow/tfjs-core/blob/52a7e4c00d30635a04da146a4652619455b0a10e/src/io/io_utils.ts#L32)):*
> decodeWeights

Good point. In TensorFlow and TensorFlow.js, non-weight tensors generally don't have names. So `decodeWeights` and `encodeWeights` are more appropriate names here. Thanks. Done.

---

*[src/io/io_utils.ts, line 34 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAtFy5WCQdvs979ljM8:-LAurpCzAeNuk_2NOPkr:b-wci4ka) ([raw file](https://github.com/tensorflow/tfjs-core/blob/52a7e4c00d30635a04da146a4652619455b0a10e/src/io/io_utils.ts#L34)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

How about calling it something more descriptive like `tensorsToBuffer()` and the other one `bufferToTensors()` or do you want to keep its implementation detail hidden from the user?
</blockquote></details>

The `encodeTensors` method actually returns two things, a buffer and an array of specs. So `tensorsToBuffers` is not the most appropriate.

Considering Ping's comment, I'll rename the methods to `encodeWeights` and `decodeWeights`.

---

*[src/io/io_utils.ts, line 34 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAtOZyY1HPG_3st-NYK:-LAus9b1AeRG4pTCaBF4:b-pcwtkt) ([raw file](https://github.com/tensorflow/tfjs-core/blob/52a7e4c00d30635a04da146a4652619455b0a10e/src/io/io_utils.ts#L34)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

Also, how different is this code from the loadWeights method, are there plan to merge those two? 
I also want to raise attention about the PR by Adam in converter repo that supports weight quantization.  
</blockquote></details>

Yes, `loadWeights` should use `decodeWeights` (formerly `decodeTensors`). I added a TODO item.

I also added TODO items to handle quantization to these methods. cc @adarob 

---

*[src/io/io_utils.ts, line 35 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAtFTgOARV79CanKciy:-LAuuYsE2c845p4JWe-D:b79o0vg) ([raw file](https://github.com/tensorflow/tfjs-core/blob/52a7e4c00d30635a04da146a4652619455b0a10e/src/io/io_utils.ts#L35)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

instead of a tuple, it's better to return an object for readability. Perhaps:
`{ buffer: ArrayBuffer, metadata: WeightsManifestEntry[]}`
</blockquote></details>

Done. I'm using the field names `data` and `specs`, to be consistent with the fields `weightData` and `weightSpecs` in `ModelArtifacts` in `type.ts`.

---

*[src/io/io_utils.ts, line 37 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAtEk41-VSpSE8OScRA:-LAuxqkgAlMfPxLz2GjK:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/52a7e4c00d30635a04da146a4652619455b0a10e/src/io/io_utils.ts#L37)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

> ```
> Float32Array|Int32Array|Uint8Array>
> ```

use the alias in `types.ts`: `TypedArray`
</blockquote></details>

Done.

---

*[src/io/io_utils.ts, line 39 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAtHwlNBploTPWRhjNd:-LAuskHe5bbIc-EKS9AU:b-er4jgq) ([raw file](https://github.com/tensorflow/tfjs-core/blob/52a7e4c00d30635a04da146a4652619455b0a10e/src/io/io_utils.ts#L39)):*
<details><summary><i>Previously, davidsoergel (David Soergel) wrote…</i></summary><blockquote>

> ```
> const tensor
> ```
This shadows the imported tensor() function; can you use a different name?
</blockquote></details>

Good point. Done. Using `t`.

---

*[src/io/io_utils.ts, line 43 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAsRw14TMG9HsRbhSxk-r1-43:-LAuw6ooEdPnDh0fRbIf:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/52a7e4c00d30635a04da146a4652619455b0a10e/src/io/io_utils.ts#L43)):*
<details><summary><i>Previously, ericdnielsen wrote…</i></summary><blockquote>

Can we include the name of the tensor here as well.
</blockquote></details>

Done.

---

*[src/io/io_utils.ts, line 72 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAtGPXl5GQfpMWDUiMQ:-LAuw88ZCbQoX7WUUBA9:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/52a7e4c00d30635a04da146a4652619455b0a10e/src/io/io_utils.ts#L72)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

util.sizeFromShape(shape)
</blockquote></details>

Done.

---

*[src/io/io_utils.ts, line 88 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAsRw14TMG9HsRbhSxk-r1-88:-LAuvkgb1p5Ob7bH8QCw:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/52a7e4c00d30635a04da146a4652619455b0a10e/src/io/io_utils.ts#L88)):*
<details><summary><i>Previously, ericdnielsen wrote…</i></summary><blockquote>

As above, please include the name fields as well.
</blockquote></details>

Done.

---

*[src/io/io_utils.ts, line 100 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAtGo7NCDQEI8AsFTMn:-LAuxs7W7G95Sb-hY-g8:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/52a7e4c00d30635a04da146a4652619455b0a10e/src/io/io_utils.ts#L100)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

use TypedArray in types.ts (alias for all these arrays)
</blockquote></details>

Done.

---

*[src/io/io_utils.ts, line 104 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAtFEkVA9iyz1b9eFeT:-LAuxsYlB6XJVCZgFVXs:b-hqjpkc) ([raw file](https://github.com/tensorflow/tfjs-core/blob/52a7e4c00d30635a04da146a4652619455b0a10e/src/io/io_utils.ts#L104)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

when do null and undefined happen? Better to throw an error in that case and let the caller figure it out
</blockquote></details>

I removed the logic for null and undefined here and throw an error as you suggested.

---

*[src/io/io_utils.ts, line 114 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAtHVs-5OGTpG2iEZU5:-LAuwBE94fSeVlozJ_GW:bhbhlge) ([raw file](https://github.com/tensorflow/tfjs-core/blob/52a7e4c00d30635a04da146a4652619455b0a10e/src/io/io_utils.ts#L114)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

> ```
> x as any
> ```
strange that you need to cast to `any`
</blockquote></details>

This is because we are testing things that shouldn't happen in TypeScript because of type checking, i.e., things that can happen only in JavaScript.

---

*[src/io/io_utils.ts, line 120 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAsRw14TMG9HsRbhSxk-r1-120:-LAut8sXBIeDsL8zXuMO:b-zh9gty) ([raw file](https://github.com/tensorflow/tfjs-core/blob/52a7e4c00d30635a04da146a4652619455b0a10e/src/io/io_utils.ts#L120)):*
<details><summary><i>Previously, ericdnielsen wrote…</i></summary><blockquote>

Are there minification concerns here with constructor.name?
</blockquote></details>

Good question. The input here is a subtype of TypedArray, i.e., a built-in type. So most likely, there will be no minification of the constructor.

---

*[src/io/io_utils_test.ts, line 22 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAtIVNxCmJ0cfMGb2tb:-LAuwJmL2AAEEYPJROSh:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/52a7e4c00d30635a04da146a4652619455b0a10e/src/io/io_utils_test.ts#L22)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

if these are exposed to the users, use public API (import from index)
</blockquote></details>

Done.

---

*[src/io/types.ts, line 83 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAtI5hhA2CJzQ6aExuK:-LAuwKyL7xcTfZWvANzs:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/52a7e4c00d30635a04da146a4652619455b0a10e/src/io/types.ts#L83)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

no need for this TODO, since you can *always* add more fields *everywhere* ;)
</blockquote></details>

Done.

---

*[src/io/types.ts, line 99 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAtLSYFEjPp38vpZY-x:-LAuwPnW0KSkEyVCZb7a:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/52a7e4c00d30635a04da146a4652619455b0a10e/src/io/types.ts#L99)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

s/resposnes/responses
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/990)*
<!-- Sent from Reviewable.io -->

nsthorat commented:
Reviewed 2 of 8 files at r2.
Review status: 2 of 6 files reviewed at latest revision, 20 unresolved discussions.

---

*[src/io/io_utils.ts, line 18 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAtE-ME1G9F1e6lfb-_:-LAxHDzj06Zl_PGsPKrB:b-bspejz) ([raw file](https://github.com/tensorflow/tfjs-core/blob/52a7e4c00d30635a04da146a4652619455b0a10e/src/io/io_utils.ts#L18)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I still see index

---

*[src/io/io_utils.ts, line 72 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAtGPXl5GQfpMWDUiMQ:-LAxI-8V2YrRnHVA7cKF:beg9356) ([raw file](https://github.com/tensorflow/tfjs-core/blob/52a7e4c00d30635a04da146a4652619455b0a10e/src/io/io_utils.ts#L72)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

can you just use size instead of numel

---

*[src/io/io_utils.ts, line 27 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAxIolS-TswtRatzK1u:-LAxIolS-TswtRatzK1v:bwlqxb) ([raw file](https://github.com/tensorflow/tfjs-core/blob/748daa6f2ff951c82ac223c03d8fc67058c29054/src/io/io_utils.ts#L27)):*
> ```XML
> /**
>  * Encode a map from names to weight values as an ArrayBuffer, along with an
>  * `Array` of `WeightsManifestEntry` as specification of the encoded weights.
> ```

would be good to mention how this is supposed to be symmetric to our other functionality. Would also be good to say this does not shard.

---

*[src/io/io_utils.ts, line 80 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAxI7GF1qzsvG3v2jLf:-LAxI7GF1qzsvG3v2jLg:b-ojwowr) ([raw file](https://github.com/tensorflow/tfjs-core/blob/748daa6f2ff951c82ac223c03d8fc67058c29054/src/io/io_utils.ts#L80)):*
> ```XML
>     let value: Tensor;
>     if (dtype === 'float32') {
>       bytes = numel * 4;
> ```

we have a map you can use:
https://github.com/tensorflow/tfjs-core/blob/master/src/weights_loader.ts#L33

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/990)*
<!-- Sent from Reviewable.io -->

caisq commented:
Review status: all files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[src/io/io_utils.ts, line 18 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAtE-ME1G9F1e6lfb-_:-LAxgrd_4mcNKT1bIzVp:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/52a7e4c00d30635a04da146a4652619455b0a10e/src/io/io_utils.ts#L18)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

I still see index
</blockquote></details>

Done.

---

*[src/io/io_utils.ts, line 72 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAtGPXl5GQfpMWDUiMQ:-LAxfgvEB0PuNEi8nIw2:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/52a7e4c00d30635a04da146a4652619455b0a10e/src/io/io_utils.ts#L72)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

can you just use size instead of numel
</blockquote></details>

Done.

---

*[src/io/io_utils.ts, line 27 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAxIolS-TswtRatzK1u:-LAxgvKi6GJZI8qTa1Kp:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/748daa6f2ff951c82ac223c03d8fc67058c29054/src/io/io_utils.ts#L27)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

would be good to mention how this is supposed to be symmetric to our other functionality. Would also be good to say this does not shard.
</blockquote></details>

Done.

---

*[src/io/io_utils.ts, line 107 at r2](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAxVSJFBnrKkvlxokGS:-LAxgJ6F86kW2Nep43yz:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/748daa6f2ff951c82ac223c03d8fc67058c29054/src/io/io_utils.ts#L107)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

use forEach since it's a list. This way the tsc won't need to compile this into es5 compatible code.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/990)*
<!-- Sent from Reviewable.io -->

dsmilkov commented:
Review status: 0 of 7 files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[src/index.ts, line 67 at r3](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAydSzJEQJfNzaRmxNN:-LAydSzJEQJfNzaRmxNO:b-ldp6s2) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a0d14025c893c55dc1e5503f1d60cf4313394f32/src/index.ts#L67)):*
> ```XML
> 
> // Second level exports.
> export {environment, test_util, util};
> ```

move io here, so we have all second level exports together

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/990)*
<!-- Sent from Reviewable.io -->

caisq commented:
Review status: 0 of 7 files reviewed at latest revision, 17 unresolved discussions, all commit checks successful.

---

*[src/index.ts, line 67 at r3](https://reviewable.io/reviews/tensorflow/tfjs-core/990#-LAydSzJEQJfNzaRmxNN:-LAydy8uDilybnXqMYPz:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/a0d14025c893c55dc1e5503f1d60cf4313394f32/src/index.ts#L67)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

move io here, so we have all second level exports together
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/990)*
<!-- Sent from Reviewable.io -->
